src/libs/print_settings.c

```diff
@@ -1390,33 +1392,40 @@ static GList* _get_profiles()
 static void _new_printer_callback(dt_printer_info_t *printer,
                                   void *user_data)
 {
-  static int count = 0;
   const dt_lib_module_t *self = (dt_lib_module_t *)user_data;
-  const dt_lib_print_settings_t *d = (dt_lib_print_settings_t*)self->data;
-
-  char *default_printer = dt_conf_get_string("plugins/print/print/printer");
-
-  g_signal_handlers_block_by_func(G_OBJECT(d->printers),
-                                  G_CALLBACK(_printer_changed), NULL);
+  dt_lib_print_settings_t *d = (dt_lib_print_settings_t*)self->data;
 
-  dt_bauhaus_combobox_add(d->printers, printer->name);
-
-  if(!g_strcmp0(default_printer, printer->name) || default_printer[0]=='\0')
-  {
-    dt_bauhaus_combobox_set(d->printers, count);
-    _set_printer(self, printer->name);
-  }
-  count++;
-  g_free(default_printer);
-
-  g_signal_handlers_unblock_by_func(G_OBJECT(d->printers),
-                                    G_CALLBACK(_printer_changed), NULL);
+  d->printer_list = g_list_append(d->printer_list, g_strdup(printer->name));
 }
 
 void view_enter(struct dt_lib_module_t *self,
                 struct dt_view_t *old_view,
                 struct dt_view_t *new_view)
 {
+  dt_lib_print_settings_t *d = (dt_lib_print_settings_t*)self->data;
+
+  if(d->printer_list != NULL)
+  {
+    // The printer list was filled by dt_printers_discovery() in a background job.
+    // Now we fill the printer combo with the found printers.
+    char *default_printer = dt_conf_get_string("plugins/print/print/printer");
+
+    for(const GList *iter = d->printer_list; iter; iter = g_list_next(iter))
+    {
+      const gchar *printer_name = iter->data;
+      dt_bauhaus_combobox_add(d->printers, printer_name);
+    }
+
+    if(default_printer[0]=='\0')
+      dt_bauhaus_combobox_set(d->printers, 0);
+    else
+      dt_bauhaus_combobox_set_from_text(d->printers, default_printer);
+
+    g_free(default_printer);
+    g_list_free_full(d->printer_list, g_free);
```

Comment:
Is this thread-safe? My bet it isn’t. Why free this list at all?

Reply:
Best would be to guard this list with mutex. Otherwise you have to think very hard if what you’ve written is actually safe and has defined behavior.

Reply:
I had in mind that when #15575 will be implemented then this list gets new entries in the background and when the print module is opened the next time the new printer will be appended to the printer combo. All other entries are left unchanged.

Cleaning the combo everytime may be a bad idea because it would then trigger a new `_set_printer()`, probably resetting all the settings the user has made.

Reply:
I have now wrapped the `printer_list` accesses with a mutex.

Reply:
Ideally you should move the mutex outside the `if(d->printer_list != NULL)`, since you read the variable in the condition. Failing to do so won’t result in crashes, but you may not get the filled list, but old value (NULL), on architectures with weak memory ordering, like ARM.